packages/opentelemetry-plugin-http/src/http.ts

```diff
@@ -0,0 +1,437 @@
+      // set the trace context header before the returned ClientRequest is
+      // ended. The Node.js docs state that the only differences between
+      // request and get are that (1) get defaults to the HTTP GET method and
+      // (2) the returned request object is ended immediately. The former is
+      // already true (at least in supported Node versions up to v10), so we
+      // simply follow the latter. Ref:
+      // https://nodejs.org/dist/latest/docs/api/http.html#http_http_get_options_callback
+      // https://github.com/googleapis/cloud-trace-nodejs/blob/master/src/plugins/plugin-http.ts#L198
+      return function outgoingGetRequest(options: string | RequestOptions | URL, callback?: HttpCallback) {
+        const req = request(options, callback);
+        req.end();
+        return req;
+      };
+    };
+  }
+
+  /**
+   * Injects span's context to header for distributed tracing and finshes the
+   * span when the response is finished.
+   * @param original The original patched function.
+   * @param options The arguments to the original function.
+   */
+  private getMakeRequestTraceFunction(
+    request: ClientRequest,
+    options: RequestOptions,
+    plugin: HttpPlugin
+  ): Func<ClientRequest> {
+    return (span: Span): ClientRequest => {
+      plugin._logger.debug('makeRequestTrace');
+
+      if (!span) {
+        plugin._logger.debug('makeRequestTrace span is null');
+        return request;
+      }
+
+      const propagation = plugin._tracer.getHttpTextFormat();
+      // If outgoing request headers contain the "Expect" header, the returned
+      // ClientRequest will throw an error if any new headers are added.
+      // So we need to clone the options object to be able to inject new
+      // header.
+      if (HttpPlugin.hasExpectHeader(options)) {
+        options = Object.assign({}, options);
+        options.headers = Object.assign({}, options.headers);
+      }
+      propagation.inject(span.context(), HttpPlugin.PROPAGATION_FORMAT, options.headers);
+
+      request.on('response', (response: IncomingMessage) => {
+        plugin._tracer.scopeManager.bind(response);
+        plugin._logger.debug('outgoingRequest on response()');
+        response.on('end', () => {
+          plugin._logger.debug('outgoingRequest on end()');
+          const method = response.method ? response.method.toUpperCase() : 'GET';
+          const headers = options.headers;
+          const userAgent = headers ? headers['user-agent'] || headers['User-Agent'] : null;
+
+          const host = options.hostname || options.host || 'localhost';
+          span
+            .setAttribute(HttpPlugin.ATTRIBUTE_HTTP_HOST, host)
+            .setAttribute(HttpPlugin.ATTRIBUTE_HTTP_METHOD, method);
+          if (options.path) {
+            span
+              .setAttribute(HttpPlugin.ATTRIBUTE_HTTP_PATH, options.path)
+              .setAttribute(HttpPlugin.ATTRIBUTE_HTTP_ROUTE, options.path);
```

Comment:
Just curious to know, `ATTRIBUTE_HTTP_PATH` vs  `ATTRIBUTE_HTTP_ROUTE`?

Reply:
path is just the normal path whereas route exclude params in path
ATTRIBUTE_HTTP_PATH = /test/321434323
ATTRIBUTE_HTTP_ROUTE = /test/:id

but here options.path in both is not correct.

I think that we should expose some feature like this : https://www.npmjs.com/package/express-prom-bundle#example-1-add-custom-regexp
it would be awesome to normalize operation list in tracer

Reply:
We should not set the route here then since the http server don't have the information

Reply:
Yeah, I think we would need to instrument the actual router (maybe express can do that) to get the matched route